tutorial/source/enumeration.ipynb

```diff
@@ -0,0 +1,516 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Running model with 10 data points\n",
+      "x.shape = torch.Size([10])\n",
+      "dist.Normal(loc[x], scale).batch_shape = torch.Size([10])\n",
+      "Running model with 10 data points\n",
+      "x.shape = torch.Size([3, 1])\n",
+      "dist.Normal(loc[x], scale).batch_shape = torch.Size([3, 1])\n"
+     ]
+    }
+   ],
+   "source": [
+    "@config_enumerate(default=\"parallel\")\n",
+    "def model(data, num_components=3):\n",
+    "    print('Running model with {} data points'.format(len(data)))\n",
+    "    p = pyro.sample(\"p\", dist.Dirichlet(0.5 * torch.ones(3)))\n",
+    "    scale = pyro.sample(\"scale\", dist.LogNormal(0, num_components))\n",
+    "    with pyro.plate(\"components\", num_components):\n",
+    "        loc = pyro.sample(\"loc\", dist.Normal(0, 10))\n",
+    "    with pyro.plate(\"data\", len(data)):\n",
+    "        x = pyro.sample(\"x\", dist.Categorical(p))\n",
+    "        print(\"x.shape = {}\".format(x.shape))\n",
+    "        pyro.sample(\"obs\", dist.Normal(loc[x], scale), obs=data)\n",
+    "        print(\"dist.Normal(loc[x], scale).batch_shape = {}\".format(\n",
+    "            dist.Normal(loc[x], scale).batch_shape))\n",
+    "        \n",
+    "guide = AutoDiagonalNormal(poutine.block(model, hide=[\"x\", \"data\"]))\n",
+    "\n",
+    "data = torch.randn(10)\n",
+    "        \n",
+    "pyro.clear_param_store()\n",
+    "elbo = TraceEnum_ELBO(max_plate_nesting=1)\n",
+    "elbo.loss(model, guide, data);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Observe that the model is run twice, first by the `AutoDiagonalNormal` to trace sample sites, and second by `elbo` to compute loss. In the first run, `x` has the standard interpretation of one sample per datum, hence shape `(10,)`. In the second run enumeration can use the same three values `(3,1)` for all data points, and relies on broadcasting for any dependent sample or observe sites that depend on data. For example, in the `pyro.sample(\"obs\",...)` statement, the distribution has shape `(3,1)`, the data has shape`(10,)`, and the broadcasted log probability tensor has shape `(3,10)`.\n",
+    "\n",
+    "For more in-depth treatment of enumeration in mixture models, see the [Gaussian Mixture Model Tutorial](http://pyro.ai/examples/gmm.html) and the.\n",
+    "\n",
+    "### Dependencies among plates <a class=\"anchor\" id=\"Dependencies-among-plates\"></a>\n",
+    "\n",
+    "The computational savings of enumerating in vectorized plates comes with restrictions on the dependency structure of models. These restrictions are in addition to the usual restrictions of conditional independence. Fore completeness we list all three restrictions:\n",
```

Comment:
typo 'Fore'

Reply:
re: "all three restrictions"

Note that Pyro includes correctness checks for all three restrictions. (right?)